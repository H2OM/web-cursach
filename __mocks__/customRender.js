import ClientContext from '@/lib/context/ClientContext';
import { render } from '@testing-library/react';

export default customRender = (ui, {providerProps}) => {
    return render(
      <ClientContext.Provider value={{...providerProps}}>{ui}</ClientContext.Provider>,
    )
}