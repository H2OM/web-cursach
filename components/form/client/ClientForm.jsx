'use client';

import ClientContext from "@/lib/context/ClientContext";
import MaskInput from "@/lib/inputMask/InputMask";
import { useContext, useEffect, useRef, useState } from "react";

export default function ClientForm () {
    const {setNotification} = useContext(ClientContext);
    const [isSubmiting, setSubmiting] = useState(false);
    const [nameInput, setNameInput] = useState('');
    const [check, setCheck] = useState(false);
    const mailRef = useRef(null);
    const phoneRef = useRef(null);
    const [errors, setErrors] = useState({
        mail: false,
        phone: false,
        name: false,
        agreement: false
    });

    const formSubmitHandler = async (e) =>{
        e.preventDefault();
        let error = false;
        if (!check) {
            if(!errors.agreement) setErrors((prev)=>({...prev, agreement: true}));
            error = true;
        }
        if(RegExp(/_/).test(phoneRef.current.value)) {
            if(!errors.phone) setErrors((prev)=>({...prev, phone: true}));
            error = true;
        }
        if(mailRef.current.value != '' && mailRef.current.value.match(/^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$/g) === null) {
            if(!errors.mail) setErrors((prev)=>({...prev, mail: true}));
            error = true;
        }
        if(nameInput.length < 3 || nameInput.length > 70) {
            if(!errors.name) setErrors((prev)=>({...prev, name: true}));
            error = true;
        }
        if(errors.name && nameInput.length >= 3 && nameInput <= 70) {
            setErrors((prev)=>({...prev, name: false}));
        }
        if(error) {return;}
        setSubmiting(true);
        const formData = new FormData(e.target);
        await fetch('/api/form/form-callback', {method: 'POST',  body: formData})
            .then((data)=>{
                if(!data.ok) {
                    setNotification((current)=>({...current, status: false}));
                }
                return data.json();
            })
            .then((respons)=>{
                if(!respons.message) {throw new Error();}
                setNotification((current)=>({...current, message: respons.message}));
                e.target.reset();
                setNameInput('');
            })
            .catch(()=>{
                setNotification((current)=>({...current, status: false, message: 'Непредвиденная ошибка, попробуйте перезагрузить страницу.'}));
            })
            .finally(()=>setSubmiting(false));
    }

   
    useEffect(()=>{
        if(nameInput.match(/^\s+|.*\s{3,}/g) !== null) {
            setNameInput((prev)=> prev.replace(/^\s+/g, '').replace(/\s{3,}/g, '  '));
        }
        if(nameInput.match(/([А-Яа-я]+[ ]{0,})+/g) === null) setNameInput('');
        else if(nameInput.match(/([А-Яа-я]+[ ]{0,})+/g) !== null && nameInput.match(/([А-Яа-я]+[ ]{0,})+/g)[0].length != nameInput.length) {
            setNameInput((prev)=>prev.match(/([А-Яа-я]+[ ]{0,})+/g).join(''));
        }
    }, [nameInput]);

    return (
        <form className={"Form__form" + (isSubmiting ? " Form__form_submit" : "")} onSubmit={formSubmitHandler}> 
            {isSubmiting ? <div className="Form__form__blur"></div> : null}
            <label role="error" htmlFor="name" className={"Form__form__input__error " + (errors.name ? "Form__form__input__error_show" : "")}>Неверный формат имени</label>
            <input className="Form__form__input" value={nameInput} name="name" type="text" 
                minLength={3} maxLength={70} placeholder="Ваше имя" required
                onChange={(e)=>{
                    e.preventDefault();
                    setNameInput(e.target.value);
                    if(errors.name && nameInput.length >= 3 && nameInput.length <= 70) {
                        setErrors((prev)=>({...prev, name: false}));
                    }
                }}/>
            
            <label role="error" htmlFor="email" className={"Form__form__input__error " + (errors.mail ? "Form__form__input__error_show" : "")}>Неверный формат почты</label>
            <input ref={mailRef} className="Form__form__input" name="mail" type="email" placeholder="Ваша почта" 
                onChange={()=>{
                    if(errors.mail && mailRef.current.value.match(/^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$/g) !== null) {
                        setErrors((prev)=>({...prev, mail: false}));
                    }
                }}/>
            <label role="error" htmlFor="number" className={"Form__form__input__error " + (errors.phone ? "Form__form__input__error_show" : "")}>Неверный номер телефона</label>
            <MaskInput inputRef={phoneRef} error={errors.phone} setError={setErrors} className={"Form__form__input"} name={"number"} placeholder={"Номер телефона"} always={false}/>
            <textarea className="Form__form__input Form__form__input_area" name="question" cols="30" rows="10" minLength={10} maxLength={1200} placeholder="Какой у вас вопрос?" required/>
            <div className="Form__form__radio">
                <label htmlFor="agreement" className={"Form__form__radio__label " + (errors.agreement ? "Form__form__radio__label_error" : "")}>Подтверждение обработки данных</label>
                <input type="checkbox" name="agreement" className="Form__form__input Form__form__input_radio"
                checked={check} onChange={({target})=>{
                    setCheck(target.value);
                    if(target.value) setErrors((prev)=>({...prev, agreement: false}));                        
                }}/>
            </div>
            <button className={"Form__form__submit" + (isSubmiting ? " Form__form__submit_submiting" : "")} type="submit" 
                disabled={isSubmiting}>
                    Отправить
            </button>
        </form>
    )
}