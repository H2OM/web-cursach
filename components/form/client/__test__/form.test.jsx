import { render, screen } from "@testing-library/react";
import ClientForm from "../ClientForm";
import customRender from "@/__mocks__/customRender";
import userEvent from "@testing-library/user-event";
import settingFetchResult from "@/__mocks__/fetchMock";
const notifications = {
    setNotification: ()=>{},
    notification: {
        status: false,
        message: ""
    }
}

describe('test form', ()=>{
    settingFetchResult({message:"done"});
    
    it('test render', async ()=>{
        const {container} = customRender(<ClientForm/>, {providerProps: notifications});
        const errors = screen.getAllByRole('error');
        const inputs = screen.getAllByPlaceholderText(/(Ваше имя){1}|(Ваша почта){1}|(Какой у вас вопрос?){1}|(Номер телефона){1}/);
       
        expect(errors.length).toBe(3);
        expect(inputs.length).toBe(4);
        expect(screen.getByRole('checkbox')).toBeInTheDocument();
        expect(screen.getByRole('button')).toBeInTheDocument();
    
        expect(container).toMatchSnapshot();
    
    });
    it('functional test - success way',async ()=>{
        customRender(<ClientForm/>, {providerProps: notifications});
        const name = screen.getByPlaceholderText('Ваше имя');
        const mail = screen.getByPlaceholderText('Ваша почта');
        const number = screen.getByPlaceholderText('Номер телефона');
        const text = screen.getByPlaceholderText('Какой у вас вопрос?');
        const submit = screen.getByRole('button');
        const checkbox = screen.getByRole('checkbox');

        await userEvent.type(name, 'Имя');
        await userEvent.type(mail, 'example@mail.ru');
        await userEvent.type(text, "some question");
        await userEvent.type(number, "79182195584");

        expect(name).toHaveValue("Имя");
        expect(number).toHaveValue("+7 (918) 219-55-84");

        await userEvent.click(checkbox);
        await userEvent.click(submit);
        
        expect(name).toHaveValue("");
        expect(number).toHaveValue("");
    });
    it('functional test - rejected due to checkbox',async ()=>{
        customRender(<ClientForm/>, {providerProps: notifications});
        const name = screen.getByPlaceholderText('Ваше имя');
        const mail = screen.getByPlaceholderText('Ваша почта');
        const text = screen.getByPlaceholderText('Какой у вас вопрос?');
        const submit = screen.getByRole('button');

        await userEvent.type(name, 'Имя');
        await userEvent.type(mail, 'example@mail.ru');
        await userEvent.type(text, "some question");

        await userEvent.click(submit);
        
        expect(name).toHaveValue("Имя");
    });
    it('functional test - rejected due to wrong mail',async ()=>{
        
        customRender(<ClientForm/>, {providerProps: notifications});
        const name = screen.getByPlaceholderText('Ваше имя');
        const mail = screen.getByPlaceholderText('Ваша почта');
        const text = screen.getByPlaceholderText('Какой у вас вопрос?');
        const submit = screen.getByRole('button');
        const mailerror = screen.getAllByRole('error')[1];
        const checkbox = screen.getByRole('checkbox');
        
        await userEvent.type(name, 'Имя');
        await userEvent.type(mail, 'example@mailru');
        await userEvent.type(text, "some question");

        await userEvent.click(checkbox);
        await userEvent.click(submit);
        expect(mail).toHaveValue("example@mailru");
        expect(mailerror.className.includes('Form__form__input__error_show')).toBeTruthy();

        await userEvent.clear(mail);
        await userEvent.type(mail, 'example@mail.ru');
        expect(mailerror.className.includes('Form__form__input__error_show')).toBeFalsy();

        await userEvent.click(submit);
        expect(mail).toHaveValue("");
        
    });
    it('functional test - test wrong name',async ()=>{
        customRender(<ClientForm/>, {providerProps: notifications});
        const name = screen.getByPlaceholderText('Ваше имя');

        await userEvent.type(name, 'Имя4223');
        expect(name).toHaveValue("Имя");
        
    });
    
});
