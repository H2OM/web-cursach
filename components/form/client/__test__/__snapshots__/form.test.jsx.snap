// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test form test render 1`] = `
<div>
  <form
    class="Form__form"
  >
    <label
      class="Form__form__input__error "
      for="name"
      role="error"
    >
      Неверный формат имени
    </label>
    <input
      class="Form__form__input"
      maxlength="70"
      minlength="3"
      name="name"
      placeholder="Ваше имя"
      required=""
      type="text"
      value=""
    />
    <label
      class="Form__form__input__error "
      for="email"
      role="error"
    >
      Неверный формат почты
    </label>
    <input
      class="Form__form__input"
      name="mail"
      placeholder="Ваша почта"
      type="email"
    />
    <label
      class="Form__form__input__error "
      for="number"
      role="error"
    >
      Неверный номер телефона
    </label>
    <input
      class="Form__form__input"
      data-target="имя"
      name="number"
      placeholder="Номер телефона"
      type="text"
    />
    <textarea
      class="Form__form__input Form__form__input_area"
      cols="30"
      maxlength="1200"
      minlength="10"
      name="question"
      placeholder="Какой у вас вопрос?"
      required=""
      rows="10"
    />
    <div
      class="Form__form__radio"
    >
      <label
        class="Form__form__radio__label "
        for="agreement"
      >
        Подтверждение обработки данных
      </label>
      <input
        class="Form__form__input Form__form__input_radio"
        name="agreement"
        type="checkbox"
      />
    </div>
    <button
      class="Form__form__submit"
      type="submit"
    >
      Отправить
    </button>
  </form>
</div>
`;
